spinta/datasets/keymaps/sqlalchemy.py

```diff
@@ -57,31 +57,48 @@ def encode(self, name: str, value: object) -> Optional[str]:
             if value is None:
                 return None
             value = _encode_value(value)
-
-        # Get value hash.
+        table = self.get_table(name)
         value = msgpack.dumps(value, strict_types=True)
         hash = hashlib.sha1(value).hexdigest()
-
-        # Try to find key by value hash.
-        table = self.get_table(name)
         query = sa.select([table.c.key]).where(table.c.hash == hash)
         key = self.conn.execute(query).scalar()
-        if key is None:
+        if primary_key is not None:
+            parent_table = name.split(".")
+            _query_parent_table = sa.select((self.metadata.tables[parent_table[0]]))
+            result = self.conn.execute(_query_parent_table)
+            for n in result:
```

Comment:
Use [one_or_none](https://docs.sqlalchemy.org/en/14/core/connections.html#sqlalchemy.engine.Result.one_or_none) instead. And query pkey by value hash, instead of looping through all keys.